prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -3528,6 +3537,73 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
         Ok(stmts)
     }
 
+    /// Collect all the available spans of the items of a contract 
+    /// and store them. Purpose of this function is to later pass these
+    /// spans to prusti-assistant, so users can look up contracts of 
+    /// function calls
+    fn store_contract_spans(
+        &self,
+        called_def_id: ProcedureDefId,
+        contract: &ProcedureContract<'tcx>,
+        call_site_span: Span,
+        substs: ty::subst::SubstsRef<'tcx>,
+    ) {
+        let mut precondition_spans:Vec<Span> = contract
+            .functional_precondition(self.encoder.env(), substs)
+            .iter()
+            .map(|(ts,_)| self.encoder.env().query.get_def_span(ts))
+            .collect();
+        let mut postcondition_spans: Vec<Span> = contract
+            .functional_postcondition(self.encoder.env(), substs)
+            .iter()
+            .map(|(ts,_)| self.encoder.env().query.get_def_span(ts))
+            .collect();
+        let purity: Option<Span> = contract
+            .specification.purity
+            .extract_with_selective_replacement()
+            .copied().unwrap_or(None)
+            .map(|id| self.encoder.env().query.get_def_span(id));
+        let terminates: Option<Span> = contract.specification.terminates
+            .extract_with_selective_replacement()
+            .copied().unwrap_or(None)
+            .map(|id| self.encoder.env().query.get_def_span(id));
+
+        // for each pledge, if it has a lefthandside, join the 2, otherwise
+        // just return span of the righthandside
+        let mut pledges: Vec<Span> = contract.pledges()
+            .map(|pledge| {
+                if let Some(lhs_defid) = pledge.lhs {
+                    let lhs_span = self.encoder.env().query.get_def_span(lhs_defid);
+                    let rhs_span = self.encoder.env().query.get_def_span(pledge.rhs);
+                    lhs_span.to(rhs_span)
+                } else {
+                    self.encoder.env().query.get_def_span(pledge.rhs)
+                }
+            })
+            .collect();
+        let mut result = vec![]; 
```

Comment:
It looks to me like you have a bunch of `DefId`s, for each of them you end up calling `get_def_span`. How about: create an iterator chain of all of the `DefId`s, then `map` `get_def_span` over it once, then `collect` into a `Vec` rather than append to it like in the following lines…

Reply:
I somewhat adjusted it, but it's still not very pretty.
I tried to generate a chain of iterators, but I did not manage to make this prettier at all. 
For the pledges I still have to merge them after finding the spans, so I can not include them. For the others it's also not quite as simple as it might look since there are 2 vecs and 2 `Option<Option<DefId>>`. I put them all into a vector now before querying the spans. I hope that's alright.